File: BluetoothLightLib/src/main/java/com/telink/util/EventBus.java
/*
 * Copyright (C) 2015 The Telink Bluetooth Light Project
 *
 */
package com.telink.util;

import android.os.Handler;
import android.os.Looper;

import com.telink.bluetooth.TelinkLog;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

public class EventBus<T> {

    private static final int CPU_COUNT = Runtime.getRuntime().availableProcessors();
    private static final int CORE_POOL_SIZE = CPU_COUNT + 1;
    private static final int MAXIMUM_POOL_SIZE = CPU_COUNT * 2 + 1;
    private static final int KEEP_ALIVE = 1;
    private static final BlockingQueue<Runnable> sPoolWorkQueue =
            new LinkedBlockingQueue<>(128);
    private static final ThreadFactory sThreadFactory = new DefaultThreadFactory();

    private static final ExecutorService EXECUTOR_SERVICE = new ThreadPoolExecutor(CORE_POOL_SIZE, MAXIMUM_POOL_SIZE, KEEP_ALIVE,
            TimeUnit.SECONDS, sPoolWorkQueue, sThreadFactory); //Executors.newCachedThreadPool(new DefaultThreadFactory());

    protected final Map<T, List<EventListener<T>>> mEventListeners = new ConcurrentHashMap<>();
    protected final Queue<Event<T>> mEventQueue = new ConcurrentLinkedQueue<>();
    protected final Handler mCurrentThreadHandler = new Handler(Looper.myLooper());
    protected final Handler mMainThreadHandler = new Handler(Looper.getMainLooper());
    private final Object mLock = new Object();
    protected boolean processing = false;
    private final Runnable task = new Runnable() {
        @Override
        public void run() {
            processEvent();
        }
    };

    public void addEventListener(T eventType, EventListener<T> listener) {

        synchronized (this.mEventListeners) {
            List<EventListener<T>> listeners;

            if (this.mEventListeners.containsKey(eventType)) {
                listeners = this.mEventListeners.get(eventType);
            } else {
                listeners = new ArrayList<>();
                this.mEventListeners.put(eventType, listeners);
            }

            if (!listeners.contains(listener)) {
                listeners.add(listener);
            }
        }
    }

    public void removeEventListener(EventListener<T> listener) {
        synchronized (this.mEventListeners) {
            for (T eventType : this.mEventListeners.keySet()) {
                this.removeEventListener(eventType, listener);
            }
        }
    }

    public void removeEventListener(T eventType, EventListener<T> listener) {
        synchronized (this.mEventListeners) {
            if (this.mEventListeners.containsKey(eventType)) {
                List<EventListener<T>> listeners = this.mEventListeners.get(eventType);
                listeners.remove(listener);
            }
        }
    }

    public void removeEventListeners() {
        synchronized (this.mEventListeners) {
            for (T eventType : this.mEventListeners.keySet()) {
                List<EventListener<T>> listeners = this.mEventListeners.get(eventType);
                listeners.clear();
                this.mEventListeners.remove(eventType);
            }
        }
    }

    public void dispatchEvent(final Event<T> event) {
//        TelinkLog.d("event looper : " + event.getThreadMode());

        this.mEventQueue.add(event);

        TelinkLog.d("postCommand event : " + event.getType() + "--" + event.getClass().getName());

        synchronized (this.mLock) {
            if (!this.processing)
                this.processOnThread();
        }
    }

    private void processOnThread() {

        final Event<T> event;

        synchronized (mEventQueue) {
            event = this.mEventQueue.peek();
            if (event == null)
                return;
        }

        switch (event.getThreadMode()) {
            case Background:
                EXECUTOR_SERVICE.execute(task);
                break;
            case Main:
                mMainThreadHandler.post(task);
                break;
            case Default:
                mCurrentThreadHandler.post(task);
                break;
        }
    }

    private void processEvent() {
//        TelinkLog.d("process on thread : " + Thread.currentThread().getName());

        final Event<T> event;

        synchronized (mEventQueue) {
            event = mEventQueue.poll();
            if (event == null)
                return;
        }

//        TelinkLog.d("process postCommand event : " + event.getType() + "--" + event.getClass().getName());

        T eventType = event.getType();
        List<EventListener<T>> listeners = null;

        synchronized (this.mEventListeners) {
            if (this.mEventListeners.containsKey(eventType)) {
                listeners = this.mEventListeners.get(eventType);
            }
        }

        if (listeners != null && !listeners.isEmpty()) {
            synchronized (this.mLock) {
                this.processing = true;
            }

            for (EventListener<T> listener : listeners) {
                if (listener != null)
                    listener.performed(event);
            }
        }

        this.processEventCompleted();
    }

    private void processEventCompleted() {
        synchronized (this.mLock) {
            this.processing = false;
        }

        if (!this.mEventQueue.isEmpty())
            this.processOnThread();
    }

    private static class DefaultThreadFactory implements ThreadFactory {
        private static final AtomicInteger POOL_NUMBER = new AtomicInteger(1);
        private final AtomicInteger threadNumber = new AtomicInteger(1);
        private final ThreadGroup group;
        private final String namePrefix;

        DefaultThreadFactory() {
            SecurityManager s = System.getSecurityManager();
            group = (s != null) ? s.getThreadGroup() :
                    Thread.currentThread().getThreadGroup();
            namePrefix = "pool-" +
                    POOL_NUMBER.getAndIncrement() +
                    "-thread-";
        }

        @Override
        public Thread newThread( final Runnable r) {

            Runnable run = new Runnable() {
                @Override
                public void run() {
                    android.os.Process.setThreadPriority(android.os.Process.THREAD_PRIORITY_BACKGROUND);
                    r.run();
                }
            };

            Thread thread = new Thread(group, run,
                    namePrefix + threadNumber.getAndIncrement(),
                    0);
            if (thread.isDaemon())
                thread.setDaemon(false);
            if (thread.getPriority() != Thread.NORM_PRIORITY)
                thread.setPriority(Thread.NORM_PRIORITY);

            return thread;
        }
    }
}

File: BluetoothLightLib/src/main/java/com/telink/event/EventBus.kt
package com.telink.event

import android.os.Handler
import android.os.Looper
import com.telink.util.BleLog
import java.util.*
import java.util.concurrent.*
import java.util.concurrent.atomic.AtomicInteger

open class EventBus<T> {

    private val mEventListeners: MutableMap<T, List<EventListener<T>>> = ConcurrentHashMap()
    private val mEventQueue: Queue<Event<T>> = ConcurrentLinkedQueue()
    private val mCurrentThreadHandler = Handler(Looper.myLooper())
    private val mMainThreadHandler = Handler(Looper.getMainLooper())
    private val mLock = Any()
    private var processing = false
    private val task = Runnable { processEvent() }

    fun addEventListener(eventType: T, listener: EventListener<T>) {

        synchronized(mEventListeners) {
            val listeners: MutableList<EventListener<T>>

            if (mEventListeners.containsKey(eventType)) {
                listeners = mEventListeners[eventType] as MutableList<EventListener<T>>
            } else {
                listeners = ArrayList()
                mEventListeners.put(eventType, listeners)
            }

            if (!listeners.contains(listener)) {
                listeners.add(listener)
            }
        }
    }

    fun removeEventListener(listener: EventListener<T>) {
        synchronized(mEventListeners) {
            for (eventType in mEventListeners.keys) {
                removeEventListener(eventType, listener)
            }
        }
    }

    fun removeEventListener(eventType: T, listener: EventListener<T>) {
        synchronized(mEventListeners) {
            if (mEventListeners.containsKey(eventType)) {
                val listeners =  mEventListeners[eventType] as MutableList<EventListener<T>>
                listeners.remove(listener)
            }
        }
    }

    fun removeEventListeners() {
        synchronized(mEventListeners) {
            for (eventType in mEventListeners.keys) {
                val listeners = mEventListeners[eventType] as MutableList<EventListener<T>>
                listeners.clear()
                mEventListeners.remove(eventType)
            }
        }
    }

    open fun dispatchEvent(event: Event<T>) {
        mEventQueue.add(event)
        BleLog.d("postCommand event : " + event.type + "--" + event.javaClass.name)
        synchronized(mLock) {
            if (!processing)
                processOnThread()
        }
    }

    private fun processOnThread() {
        var event: Event<T>? = null
        synchronized(mEventQueue) {
            event = mEventQueue.peek()
            if (event == null)
                return
        }
        when (event?.threadMode) {
            Event.ThreadMode.Background -> EXECUTOR_SERVICE.execute(task)
            Event.ThreadMode.Main -> mMainThreadHandler.post(task)
            Event.ThreadMode.Default -> mCurrentThreadHandler.post(task)
        }
    }

    private fun processEvent() {
        var event: Event<T>? = null
        synchronized(mEventQueue) {
            event = mEventQueue.poll()
            if (event == null)
                return
        }
        val eventType = event?.type
        var listeners: List<EventListener<T>>? = null

        synchronized(mEventListeners) {
            if (mEventListeners.containsKey(eventType)) {
                listeners = mEventListeners[eventType]
            }
        }

        if ( !listeners!!.isEmpty()) {
            synchronized(mLock) {
                processing = true
            }
            for (listener in listeners!!) {
                listener?.performed(event!!)
            }
        }
        processEventCompleted()
    }

    private fun processEventCompleted() {
        synchronized(mLock) {
            processing = false
        }
        if (!mEventQueue.isEmpty())
            processOnThread()
    }

    private class DefaultThreadFactory internal constructor() : ThreadFactory {
        private val threadNumber = AtomicInteger(1)
        private val group: ThreadGroup
        private val namePrefix: String

        init {
            val s = System.getSecurityManager()
            group = if (s != null)
                s.threadGroup
            else
                Thread.currentThread().threadGroup
            namePrefix = "pool-${POOL_NUMBER.getAndIncrement()}-thread-"
        }

        override fun newThread(r: Runnable): Thread {

            val run = Runnable {
                android.os.Process.setThreadPriority(android.os.Process.THREAD_PRIORITY_BACKGROUND)
                r.run()
            }

            val thread = Thread(group, run,namePrefix + threadNumber.getAndIncrement(),0)
            if (thread.isDaemon)
                thread.isDaemon = false
            if (thread.priority != Thread.NORM_PRIORITY)
                thread.priority = Thread.NORM_PRIORITY

            return thread
        }

        companion object {
            private val POOL_NUMBER = AtomicInteger(1)
        }
    }

    companion object {

        private val CPU_COUNT = Runtime.getRuntime().availableProcessors()
        private val CORE_POOL_SIZE = CPU_COUNT + 1
        private val MAXIMUM_POOL_SIZE = CPU_COUNT * 2 + 1
        private val KEEP_ALIVE = 1
        private val sPoolWorkQueue = LinkedBlockingQueue<Runnable>(128)
        private val sThreadFactory = DefaultThreadFactory()

        private val EXECUTOR_SERVICE = ThreadPoolExecutor(CORE_POOL_SIZE, MAXIMUM_POOL_SIZE, KEEP_ALIVE.toLong(),
                TimeUnit.SECONDS, sPoolWorkQueue, sThreadFactory)
    }
}
